### ISSUE TITLE
SSURGO and OSM Default

### ISSUE THREAD

**Issue by FLO-2DKaren** [FLO-2DSoftware/qgis-flo-2d-plugin#1580]
Let's make the Default NO in these.  I keep accidentally clicking yes.

https://github.com/user-attachments/assets/00bf21a3-7185-4985-828e-497f37aa2505

**Comment by rpachaly** [comment]
I solved this problem by changing the message to 

![image](https://github.com/user-attachments/assets/b1c8ca5a-5f3d-467f-a399-a92d19606e69)

However, I found a bug on the OSM. Working on it. When I finish this issue, remember to also test the OSM

**Comment by rpachaly** [comment]
What a weird error... The Land Use OSM data does not work if I have the OpenCL accelerator activated. 

So, this is not on us. I'll merge this branch.

**Comment by FLO-2DKaren** [comment]
Failed Tests on QGIS 3.40.1 when data is added to Groups.  See first and last part of video.
I think we should Save the Land Use and Soil to the project path because they are still good layers but you lose the data if you crash.  
Here's the video and a test project.
https://flo-2d.sharefile.com/d-sb96c8df487da45a5b6234ebfa0720cbc

**Comment by rpachaly** [comment]
I'm seeing a lot of differences in this new QGIS version.

**Comment by FLO-2DKaren** [comment]
Yep. That's why I'm using 3.40 to test everything.  We know the tests pass in 3.34 because you already did those.

**Comment by rpachaly** [comment]
What do you think about saving the soil layer and the land use and ALWAYS delete the intermediate calculation layers? We can get rid of that message then.

**Comment by FLO-2DKaren** [comment]
I don't have an issue with that so long as we can trace the files to the temp folder.  Sometimes when we get an error, I trace the temp files and load them to see if they are complete or not.

**Comment by FLO-2DKaren** [comment]
I just added the bug label so we don't forget this one is important.

**Comment by rpachaly** [comment]
Karen, I was running a lot of tests here and I don't think this is related to groups or any flo-2d plugin code. This seems to be related to QGIS. 

Do the following: run the OSM and SSURGO with the intermediate layers being saved. Then, just close the Compute Green-Ampt dialog and re-open it. Make sure that all fields are set correctly and click OK. It will run without any problems. 

If you look into the QGIS Crash log, it is related to a combobox. It does not make sense. I was looking on the QGIS repo and a lot of people are complaining about QGIS crashes. 

My opinion: let's keep on eye on this and double check when the next LTR is released.

**Comment by FLO-2DKaren** [comment]
OK I agree.  That's a valid approach.

Did you add the Save Layer Code?  I think its best to save them directly the gpkg.  If so I'll run the tests again.  I wanted to run them on several different projects before we close this issue.

- [ ] Self Help 
- [ ] Euro project
- [ ] Nevada project
- [ ] Random state project

Things to check

- [ ] Spread of all infil variables
- [ ] Large Project
- [ ] Processing time on fast computer vs work computer

**Comment by rpachaly** [comment]
No, I haven't done that yet. I'm studying for the Limiting Froude number webinar and probably with touch on this again on Friday, if that's ok.

**Comment by rpachaly** [comment]
I did some changes on the layout (we'll need to change the documentation, sorry). But it looks simpler now. 

![image](https://github.com/user-attachments/assets/d7ff2388-9c52-43f3-b0e4-a0814e67f9d4)

Now it always save the soil layer and the land user layer to the geopackage. The other layers can be saved based on that checkbox. 

I did a few tests here and all looked good. 

It is on branch https://github.com/FLO-2DSoftware/qgis-flo-2d-plugin/compare/ssurgo-osm-default?expand=1

**Comment by FLO-2DNoemi** [comment]
I don’t recall seeing the option to save intermediate calculation layers
before.

**Comment by FLO-2DKaren** [comment]
K Test thoroughly.

**Comment by FLO-2DKaren** [comment]
Final calculation is crashing QGIS.   Tested on Self Help Kit

https://github.com/user-attachments/assets/78a84c62-56b3-42ac-a6e9-8d33bb3f2e6b

**Comment by FLO-2DKaren** [comment]
I'm testing this again using grid instead of Comp Domain.

**Comment by FLO-2DKaren** [comment]
3.40.1 error message.
The calculator isn't crashing QGIS but it's throwing this error.

ERROR 140119.1715: Green-Ampt infiltration failed!.
Error while calculating xksat.
__________________________________________________

Error:
   TypeError: '>' not supported between instances of 'str' and 'int'

In file:
   C:\Users/Karen/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\flo2d\flo2d_tools\infiltration_tools.py

In function:
   calculate_xksat_weighted

On line 566:
^^^^^^^^^

**Comment by rpachaly** [comment]
It it done on the branch https://github.com/FLO-2DSoftware/qgis-flo-2d-plugin/tree/ssurgo-osm-default

I did not merge it.

**Comment by FLO-2DKaren** [comment]
Working on 3.34.  
Failing on 3.40 with this error message:

ERROR 140119.1715: Green-Ampt infiltration failed!.
Error while calculating xksat.
__________________________________________________

Error:
   TypeError: '>' not supported between instances of 'str' and 'int'

In file:
   C:\Users/Karen/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\flo2d\flo2d_tools\infiltration_tools.py

In function:
   calculate_xksat_weighted

On line 566:
^^^^^^^^^

**Comment by FLO-2DKaren** [comment]
If that error appears, we should exit the loop and tell the user to check their Fields.

**Comment by FLO-2DKaren** [comment]
I'm trying this:
**infil_editor_widget.py**
Line 574
            else:
                QApplication.restoreOverrideCursor()
                self.uc.show_critical(
                    "ERROR 061218.1839: Green-Ampt infiltration failed!. \nPlease check that the Green Ampt Fields are correctly assigned."
                )

**infiltration_tools.py**
Line 578
        except Exception as e:
            QApplication.restoreOverrideCursor()
            self.uc.show_error(
                "ERROR 140119.1715: Green-Ampt infiltration failed!."
                + "\nError while calculating xksat."
                + "\n__________________________________________________",
                + "\n Check that the correct fields were assigned.",
                e,
            )
            return None

**Comment by FLO-2DKaren** [comment]
@rpachaly 

Hi Robson,

On the Green Ampt Calculator, I want to modify two error checks so that they don't get stuck in a loop.  the one in infiltration tools needs a return none and i modified the message on two messages.  If you are OK with these, can you apply them and I'll merge that code once I finish my tests.

**Comment by FLO-2DKaren** [comment]
Failing to get data in California project.  This is the mult domain southwest geopackage.  Same error on the mult domain full geopackage.  Error on 3.34 and 3.40.  Maybe this won't work in California.  I don't know that any of their agencies use Green-Ampt.  We should check with HEC-RAS.

ERROR: Green-Ampt SSURGO infiltration parameters failed!.
__________________________________________________

Error:
   UnboundLocalError: cannot access local variable 'PSIF' where it is not associated with a value

In file:
   C:\Users/Karen/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\flo2d\gui\infil_editor_widget.py

In function:
   calculate_ssurgo

On line 1020:
^^^^

**Comment by FLO-2DKaren** [comment]
This is a project in Florida. 

ERROR: Green-Ampt SSURGO infiltration parameters failed!.
__________________________________________________

Error:
   QgsProcessingException: Feature (238) from “chorizon” has invalid geometry. Please fix the geometry or change the “Invalid features filtering” option for this input or globally in Processing settings.

In file:
   C:\Users/Karen/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\flo2d\gui\infil_editor_widget.py

In function:
   calculate_ssurgo

On line 1000:
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

**Comment by rpachaly** [comment]
Karen, I did the following to the code:

- Added the PSIF = 0
- Added back the deleteHoles function improved by removing only holes with area less than cell_size * cell_size 
- Added a smart code to use an average of chfrags if a soil horizon has several different coarse fragment sizes
- Added the check on the fields (we already discussed that on our meeting today)

I did some tests on the Self Help Kit and on the Multidomain project and all seemed good. I'll push the branch so we can test in more projects tomorrow.

**Comment by FLO-2DKaren** [comment]
OK Robson.  I can start testing on it right away!!!

**Comment by rpachaly** [comment]
It did not pass in some of the testing projects that I have here. Still working on it.

**Comment by rpachaly** [comment]
I did some changes in the code and all seems to be working fine in all test cases that I used. Here are the changes:

- Removed the deleteHoles function that was using a default QGIS processing function called Delete Holes. I did that because in some cases the hole is not fully inside a polygon., it can be touching multiple polygons. So, I created a function that does the whole process of filling the holes and setting the attributes as the same as the closest polygon. This fixed the stacked polygons and the wholes on the shapefiles. 
- I added an if statement on the Wilting Point calculation. First, it calculates the Wilting Point based on the percentages of sand, clay and organic matter. Then, it adjusts this Wilting Point using an empirical equation. I let it adjust the Wilting Point but if it returns a negative value, I use the first Wilting Point calculation. Since both are estimations from the percentages of sand, clay and organic matter, I don't think modification will impact significantly on the results. The reason for this modification is that it uses the log of the Wilting Point in an equation to calculate the Hydraulic Conductivity and, if it is negative, we get that math.domain error. So, with this change I don't expect to see that error anymore. 

If you have questions, let me know. It is on branch https://github.com/FLO-2DSoftware/qgis-flo-2d-plugin/pull/1638

**Comment by FLO-2DKaren** [comment]
testing complete working as designed.
